### PR TITLE
ft: publish to redis on success

### PR DESCRIPTION
**Pull request by rahulreddy**
This commits adds the publish feature, where in, on a successful metrics push
the UID of the request is pushed to the channel "utapi". Clients, for example
test suites, can subscribe to this channel to get notified when a metric push
succeeds.

### CODE REVIEW

**Comment by rahulreddy** [comment]
@ironman-machine 
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

File: src/lib/backend/Memory.js [hunk 1]
```diff
@@ -311,6 +313,18 @@ export default class Memory {
...
 
     /**
+    * publish a message to a channel, clients can subsribe to this channel
+    * to receive messages
+    * @param {string} channel - channel to push the message
+    * @param {string} message - message to be pushed to the channel
+    * @return {object} - current instance
+    */
+    publish(channel, message) {
+        this.channels.emit('utapi', message);
```
Comment (bennettbuchanan) [inline src/lib/backend/Memory.js]: Should the parameter `channel` instead be `channels`? I think this may be a typo. 😃

Comment (rahulreddy) [inline src/lib/backend/Memory.js]: In the future we may have multiple channels, so I just used the plural 😄

**Comment by rahulreddy** [comment]
@ironman-machine this time w/o typos

**Comment by rahulreddy** [comment]
@ironman-machine this time w/o typos
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

File: src/lib/backend/Memory.js (hunk 1 above)
Comment (bennettbuchanan) [inline src/lib/backend/Memory.js]: As it stands, with `memBackend`, the channel will always be `'utapi'`. To eventually accommodate multiple channels, shouldn't the `channel` parameter be passed to `this.channels.emit`?

**Comment by rahulreddy** [comment]
@ironman-machine more fixes
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine more more fixes
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine how about now
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine again
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine again
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

File: src/lib/backend/Memory.js (hunk 1 above)
Comment (rahulreddy) [inline src/lib/backend/Memory.js]: I wrote this memory backend to emulate Redis. The `ioredis` module sends two vars in the callback, and for me to access the second param I have to include the first param as well. I am also thinking of changing this channel name but you are right. That was a typo, channel should be used as param.

Comment (rahulreddy) [inline src/lib/backend/Memory.js]: Also consider this PR to be changing state and not final yet as I am trying to come up with a sync-async solution for Integration Utapi tests.

**Comment by rahulreddy** [comment]
@ironman-machine again
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine again
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine again
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine 
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by michael-zapata** [comment]
Since this is mainly for tests purposes, should we hide that behaviour behind a flag to avoid unneeded operations in a production environment?

**Comment by rahulreddy** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@michael-zapata 
`emit` is synchronous, it should not hold us back (in terms of performance) in production environment. So we can avoid customization for test/prod. This could be a feature for end-users who want to be notified on "when the metric is pushed".
I will address the duplication soon, I feel like opened the PR prematurely as it's not passing the tests yet.

**Comment by rahulreddy** [comment]
@ironman-machine
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine with increased timeout for utapi tests
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine set listener to once
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine try
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine try
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine try
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
@ironman-machine try
SCALITY_INTEGRATION_BRANCH=fix/flakyUtapiTests

**Comment by rahulreddy** [comment]
We are going to leverage some other strategy.

**Pull request closed without merging**